### PR TITLE
Multi modification

### LINKED ISSUE #3474: Multiple modifications and merging of modifications.
I think there are some corner cases in the handling of modifications where we need some explicit clarifications.
#3339 contains some examples.
Here are a few more models:
```
model Test
  model A
    replaceable Real x = 1;
  end A;

  A a(redeclare Real x, x.start = 2);
end Test;
```

```
model Test2
  model A
    replaceable Real x = 1;
  end A;

  model A1
    extends A(x.start = 2);
  end A1;

  A1 a(redeclare Real x);
end Test2;
```

```
model Test3
  model A
    replaceable Real x = 1;
  end A;

  model A1
    extends A(x.start = 1);
  end A1;

  type R = Real(start = 2);
  A1 a(redeclare R x);
end Test3;
```

I would consider all of them valid, and after resolving the modifications, I would expect:
`Test.a.x.start = 2`
`Test2.a.x.start = 2`
`Test3.a.x.start = 2`

What do other people think?

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #3474 

Note that:
- I only added two of the given test-cases, since the others were for merging of modifiers and I believe that should be clear enough (otherwise we can add another PR).
- I restricted this handling it to the cases we agree on, and added examples for that well. If we don't want that restriction it would be possible to exclude those commits.

### CODE REVIEW

File: chapters/inheritance.tex [hunk 1]
```diff
@@ -434,6 +437,11 @@ \subsection{Single Modification}\label{single-modification}
   // Ok, different attributes designated (unit, displayUnit and value)
   // identical to:
   C4 b(x(final unit = "V", displayUnit = "mV") = 5.0));
+
+  C4 c(final x, final x.unit = "V", x.displayUnit = "mV");
+  // Ok, different attributes and "final x" in itself is ok
+  // identical to (the final on unit is redundant).
```
Comment (henrikt-ma) [inline chapters/inheritance.tex:443]: Punctuation:
```suggestion
  // Ok, different attributes and "final x" in itself is ok,
  // identical to (the final on unit is redundant):
```

File: chapters/inheritance.tex [hunk 2]
```diff
@@ -434,6 +437,11 @@ \subsection{Single Modification}\label{single-modification}
   // Ok, different attributes designated (unit, displayUnit and value)
   // identical to:
   C4 b(x(final unit = "V", displayUnit = "mV") = 5.0));
+
+  C4 c(final x, final x.unit = "V", x.displayUnit = "mV");
+  // Ok, different attributes and "final x" in itself is ok
```
Comment (henrikt-ma) [inline chapters/inheritance.tex:442]: Take opportunity to reduce OK written as _Ok_:
```suggestion
  // OK, different attributes and "final x" in itself is OK,
```

**Review by henrikt-ma — COMMENTED** [review]
I abstain from reviewing the technical content, as this is an area I don't understand well enough.

**Comment by HansOlsson** [comment]
@qlambert-pro does this agree with your understanding?

**Comment by maltelenz** [comment]
> @qlambert-pro does this agree with your understanding?

FYI @qlambert-pro is on extended parental leave, so don't expect an answer from him in the short term.

File: chapters/inheritance.tex [hunk 3]
```diff
@@ -418,7 +418,10 @@ \subsection{Merging of Modifications}\label{merging-of-modifications}
 
 \subsection{Single Modification}\label{single-modification}
 
-Two arguments of a modification shall not modify the same element, attribute, or description-string.  When using qualified names the different qualified names starting with the same identifier are merged into one modifier.  If a modifier with a qualified name has the \lstinline!each! or \lstinline!final! prefix, that prefix is only seen as applied to the final part of the name.
+Two arguments of a modification shall not modify the same element, attribute, or description-string.
+When using qualified names the different qualified names starting with the same identifier are merged into one modifier.
+This merged modifier can be described as a purely syntactic rewriting to an equivalent modifier, except in the case of replaceable redeclarations without a constraining type.
```
Comment (eshmoylova) [inline chapters/inheritance.tex:423]: Reading this for the first time without knowing that there will be an example below felt like there is a hole in the explanation. Perhaps adding, a sentence that there is an explanation coming later would help.
```suggestion
This merged modifier can be described as a purely syntactic rewriting to an equivalent modifier, except in the case of replaceable redeclarations without a constraining type.
The latter is described in the example below.
```

**Review by eshmoylova — CHANGES_REQUESTED** [review]
I believe that is consistent with what we agreed on. I like the examples. Just a souple of suggestions to clarify the text.

**Comment by HansOlsson** [comment]
> I believe that is consistent with what we agreed on. I like the examples. Just a souple of suggestions to clarify the text.

That should now be done.

**Comment by HansOlsson** [comment]
> > I believe that is consistent with what we agreed on. I like the examples. Just a souple of suggestions to clarify the text.
> 
> That should now be done.

@eshmoylova can you re-review to indicate if you agree with the update

File: chapters/inheritance.tex [hunk 4]
```diff
@@ -452,6 +461,43 @@ \subsection{Single Modification}\label{single-modification}
...
+  end A;
+
+  A a(redeclare Real x, x.start = 2);
+  // Identical to A a(redeclare Real x(start=2));
+  A a(redeclare replaceable Real x constrainedby Real, x.start = 2);
+  // Identical to A a(redeclare Real x constrainedby Real(start=2));
+end Test;
+\end{lstlisting}
+
+For replaceable redeclarations without a constraining type the merging is only conceptual.
```
Comment (eshmoylova) [inline chapters/inheritance.tex:479]: This is where I typed a comment before but it disappeared. Trying again. I don't find that "only conceptual" explains enough what is going on. Can we formulate it more concretely? From the example it looks like there is still some merging going on. Is it correct then to say that if a redeclaration is also replaceable but does not include its own constranedby-clause, a constrainedby-clause is created using the class specified in the previous constrainedby-clause and the modification added to it? This topic is one of those where I forget the all the minute details if I haven't looked at it in more than a week.

**Review by eshmoylova — CHANGES_REQUESTED** [review]
One problem is that redeclarations, constrainedby-clauses, their modifications and how to merge them are not introduced until the next section. Maybe adding some links to the corresponding sections would help?

File: chapters/inheritance.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/inheritance.tex]: I have now added the forward reference and rewritten that section,  @eshmoylova

**Review by eshmoylova — APPROVED** [review]
Looks good. Thank you.
